Sources/_Helpers/URLSession+AsyncAwait.swift

```diff
@@ -0,0 +1,116 @@
+#if canImport(FoundationNetworking)
+import Foundation
+import FoundationNetworking
+
+/// A set of errors that can be returned from the
+/// polyfilled extensions on ``URLSession``
+public enum URLSessionPolyfillError: Error {
+  /// Returned when no data and no error are provided.
+  case noDataNoErrorReturned
+}
+
+/// A private helper which let's us manage the asynchronous cancellation
+/// of the returned URLSessionTasks from our polyfill implementation.
+///
+/// This is a lightly modified version of https://github.com/swift-server/async-http-client/blob/16aed40d3e30e8453e226828d59ad2e2c5fd6355/Sources/AsyncHTTPClient/AsyncAwait/HTTPClient%2Bexecute.swift#L152-L156
+/// we use this for the same reasons as listed in the linked code in that there
+/// really isn't a good way to deal with cancellation in the 'with*Continuation' functions.
+private actor URLSessionTaskCancellationHelper {
+  enum State {
+    case initialized
+    case registered(URLSessionTask)
+    case cancelled
+  }
+
+  var state: State = .initialized
+
+  init() {}
+
+  nonisolated func register(_ task: URLSessionTask) {
+    Task {
+      await actuallyRegister(task)
+    }
+  }
+
+  nonisolated func cancel() {
+    Task {
+      await actuallyCancel()
+    }
+  }
+
+  private func actuallyRegister(_ task: URLSessionTask) {
+    switch state {
+    case .registered:
+      preconditionFailure("Attempting to register another task while the current helper already has a registered task!")
+    case .cancelled:
+      preconditionFailure("Attempting to register a task while the current handler is in the cancelled state!")
+    case .initialized:
+      state = .registered(task)
+
+    }
+  }
+
+  private func actuallyCancel() {
+    switch state {
+    case let .registered(task):
+      task.cancel()
+    case .cancelled:
+      break
+    case .initialized:
+      break
+    }
+  }
+}
+
+extension URLSession {
+  public func data(for request: URLRequest, delegate: (URLSessionTaskDelegate)? = nil) async throws -> (Data, URLResponse) {
+    let helper = URLSessionTaskCancellationHelper()
+
+    return try await withTaskCancellationHandler(operation: {
+      return try await withCheckedThrowingContinuation({ continuation in
+        let task = dataTask(with: request, completionHandler: { data, response, error in
+          if let error {
+            continuation.resume(throwing: error)
+          } else if let data, let response {
+            continuation.resume(returning: (data, response))
+          } else {
+            continuation.resume(throwing: URLSessionPolyfillError.noDataNoErrorReturned)
+          }
+        })
+
+        helper.register(task)
+
+        task.resume()
```

Comment:
i _think_ this has no guaranteed ordering relative to the effective `helper` calls. i.e. the task will start running ~here even if the on-actor `actuallyCancel`/`actuallyRegister` calls haven't yet executed. does that matter?

Reply:
Yeah I had this same question when looking at the reference code. I'm not sure if it's totally avoidable, and we might end up with whatever quirks the reference has as well. 

One complication is that the `onCancel` closure is required to be synchronous so I'm not sure how to improve upon what the async http client folks have already cooked up. Any ideas?

Reply:
i think one possibility would be to merge resumption of the task into the `register` invocation (`registerAndResume(...)`?). i'm not sure there would be any issues in practice with the code as-is (cancellation now seems like it will eventually be propagated), but since swift concurrency gives no assurances about the ordering of submitted tasks, or how long it might take before they're run, it seems like there's a theoretical opening for weird behavior where the data task runs for a non-trivial amount of time before the cancel/register calls are made (or could it even run to completion before registering/canceling in some edge case?). the main benefit of moving resumption into the registration logic might be to just eliminate having to think through this sort of scenario.

again, i don't think a change is necessarily needed here... mostly this is just a question out of curiosity.